### PR TITLE
feat: add Typescript Module SDK dev

### PR DESCRIPTION
**Pull request by TomChv**
Adds a new dev module for the Typescript SDK.
Adds the reference node module.

Fixes tsconfig updator relative path to point to the correct one.
Fixes an issue with constructor calling when required args are expected.
Simplifies default package.json in template
Fixes an issue with variadic arguments not treated as optional.

Fixes DEV-3308
Fixes DEV-3345
Related to DEV-3352

### CODE REVIEW

**Comment by gerhard** [comment]
Leaving this with you @helderco. If we can get this merged next week, it will be amazing!

File: sdk/typescript/dev/src/ci.ts
Comment (helderco) [inline sdk/typescript/dev/src/ci.ts:4]: I think the engine will namespace this so it'll probably create a TypescriptSdkDevCi object, but I prefer keeping "ci" stuff to the `/ci` (from repo root) module that'll orchestrate all SDKs, engine, CLI... That's what'll go on .workflows.

In this directory, it's about local development, but also making it reusable so that it can be as simple as possible to use from the "ci" module, and should also allow running the linter on a docs snippet, for example, by providing different source files. This way Vikram or Solomon can work on an example, without having node installed locally.

File: sdk/typescript/dev/src/ci.ts
Comment (helderco) [inline sdk/typescript/dev/src/ci.ts:19]: `sync` instead of `stdout` since you're not using that output.

File: sdk/typescript/dev/src/index.ts
Comment (helderco) [inline sdk/typescript/dev/src/index.ts:22]:
```suggestion
   * Example usage: `dagger call project shell --args=/bin/sh`
```

**Review by helderco — COMMENTED** [review]
Most of the comments aren't blockers, just suggestions. I'll afford some freedom here of course, as long as it serves the purpose of our colleagues managing Typescript in the repo without needing any tools locally. Just let me know when you think it's ready for another review.

We'll want to come up with some conventions and patterns as a team, but until then we need to experiment and see what works best 🙂

**Comment by jedevc** [comment]
@TomChv is this a milestone feature? If it's on the milestone, it should block doing the release if it is not merged by then (from my POV). From what I can tell, this is just an internal dev feature, so we shouldn't block the release on this?

**Comment by TomChv** [comment]
@jedevc This PR also fixes an issue with constructor that has required arguments.
What I suggest is that I fixe all comments @helderco  said and we merge that PR, then I'll continue with a follow up PR to add more functions.

What do you think? If you prefer I can open a Pr for the constructor issue, but since devSDK is pretty much internal, it's fine to merge it IMO even if it's not perfect

File: sdk/typescript/dev/src/ci.ts
Comment (TomChv) [reply sdk/typescript/dev/src/ci.ts]: I removed CI in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936) in favor of top level functions

File: sdk/typescript/dev/src/ci.ts
Comment (TomChv) [reply sdk/typescript/dev/src/ci.ts]: I put test, build and lint as top level function but they are scoped to the TypescriptSDK project.
I need to find a good pattern so it can accept standalone typescript files to lint too since docs snippets doesn't have js project in it

File: sdk/typescript/dev/node/src/commands.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/commands.ts]: Agreed, fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/node/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/index.ts]: It's possible though, but I'm not longer doing that since [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/node/src/commands.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/commands.ts]: Fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/package.json
Comment (TomChv) [reply sdk/typescript/dev/package.json]: Fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/node/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/index.ts:56]: Why so? My function returns an error if the package do not exist, it makes the code much cleaner that way imo

File: sdk/typescript/dev/node/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/index.ts]: Fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/node/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/index.ts]: Fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/dagger.json
Comment (TomChv) [reply sdk/typescript/dev/dagger.json:1]: Fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/.envrc
Comment (TomChv) [reply sdk/typescript/dev/.envrc:1]: My bad, fixed in [4474436](https://github.com/dagger/dagger/pull/6522/commits/4474436f1e8112cf2e70b5f2e8ea0326b27ca936)

File: sdk/typescript/dev/src/ci.ts
Comment (TomChv) [reply sdk/typescript/dev/src/ci.ts]: And I'm not sure it will work without a proper eslint configuration files, I need to do some tests

File: sdk/typescript/dev/node/src/commands.ts
Comment (helderco) [inline sdk/typescript/dev/node/src/commands.ts:10]: Isn't this redundant?

File: sdk/typescript/runtime/template/package.json
Comment (helderco) [inline sdk/typescript/runtime/template/package.json:4]: 🎉

File: sdk/typescript/dev/src/index.ts
Comment (helderco) [inline sdk/typescript/dev/src/index.ts:48]: The way I do this in python is to have a `base()` function that other methods reuse, instead of doing it in the constructor. Not saying it's better, just another option.

**Review by helderco — APPROVED** [review]

File: sdk/typescript/dev/node/src/index.ts
Comment (helderco) [reply sdk/typescript/dev/node/src/index.ts:56]: What I meant is this:

```ts
@func()
withYarn(): Node {
    this.container = this.container
        .withEntrypoint(["yarn"])
        ...
    return this
}

@fun()
withNpm(): Node {
    ...
}

@func()
withPnpm(): None {
    ...
}
```

Then you don’t need to handle the error if `pkgManager` isn’t valid (thus the enum comment). The `withYarn`, etc, utils aren’t necessary since they’re not reused elsewhere in the module as they can’t be used externally.

Better DX to see as available functions rather than providing a string that matches what’s in the argument doc.

File: sdk/typescript/dev/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/src/index.ts:48]: I like the constructor option, but I might refactor it later to be more flexible around the source provided to the module.

File: sdk/typescript/dev/node/src/index.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/index.ts:56]: Ok I'll update that

File: sdk/typescript/dev/node/src/commands.ts
Comment (TomChv) [reply sdk/typescript/dev/node/src/commands.ts:10]: What do you recommand? I need to store the container somewhere to apply command on it?

The other way would be to remove `Commands` and do everything at the top level object.

Comment (helderco) [reply sdk/typescript/dev/node/src/commands.ts:10]: Oh, I guess you made it required in this PR to explicitly create a `constructor` function, otherwise no default one will be registered?

**Pull request merged**
